flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/operations/AlterTableSchemaUtil.java

```diff
@@ -0,0 +1,457 @@
+            DataType dataType = TypeConversions.fromLogicalToDataType(logicalType);
+            RelDataType relType = typeSpec.deriveType(sqlValidator, nullable);
+            Schema.UnresolvedColumn newColumn;
+            if (isPhysical) {
+                newColumn = new Schema.UnresolvedPhysicalColumn(name, dataType, getComment(column));
+                physicalFieldNamesToTypes.put(name, relType);
+            } else {
+                newColumn =
+                        new Schema.UnresolvedMetadataColumn(
+                                name,
+                                dataType,
+                                ((SqlTableColumn.SqlMetadataColumn) column)
+                                        .getMetadataAlias()
+                                        .orElse(null),
+                                ((SqlTableColumn.SqlMetadataColumn) column).isVirtual(),
+                                getComment(column));
+                metadataFieldNamesToTypes.put(name, relType);
+            }
+            columns.put(name, newColumn);
+        }
+
+        private void convertComputedColumn(SqlTableColumn.SqlComputedColumn column) {
+            String name = column.getName().getSimple();
+            Map<String, RelDataType> accessibleFieldNamesToTypes = new HashMap<>();
+            accessibleFieldNamesToTypes.putAll(physicalFieldNamesToTypes);
+            accessibleFieldNamesToTypes.putAll(metadataFieldNamesToTypes);
+
+            // computed column cannot be generated on another computed column
+            final SqlNode validatedExpr =
+                    sqlValidator.validateParameterizedExpression(
+                            column.getExpr(), accessibleFieldNamesToTypes);
+            final RelDataType validatedType = sqlValidator.getValidatedNodeType(validatedExpr);
+
+            Schema.UnresolvedColumn newColumn =
+                    new Schema.UnresolvedComputedColumn(
+                            name,
+                            new SqlCallExpression(escapeExpressions.apply(validatedExpr)),
+                            getComment(column));
+            computedFieldNamesToTypes.put(name, validatedType);
+            columns.put(name, newColumn);
+        }
+
+        private void verifyRowtimeAttribute(
+                AlterSchemaStrategy strategy,
+                SqlIdentifier eventTimeColumnName,
+                Map<String, RelDataType> allFieldsTypes) {
+            if (!watermarkSpec.isEmpty() && strategy == AlterSchemaStrategy.ADD) {
+                throw new ValidationException(
+                        String.format(
+                                "There already exists a watermark spec for column '%s' in the base table. You "
+                                        + "might want to drop it before adding a new one.",
+                                watermarkSpec.keySet()));
+            }
+            if (!eventTimeColumnName.isSimple()) {
```

Comment:
> We can support defining watermark on nested field, I think we should align the logic in
> 
> https://github.com/apache/flink/blob/34de3989a613cf7124f9e301cb8284080f4df4ac/flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/operations/MergeTableLikeUtil.java#L373

I don't think we truly support defining a watermark on the nested column. E.g.
```sql
CREATE TABLE foo(
  f0 INT,
  f1 STRING,
  f2 ROW<f21 STRING, f22 TIMESTAMP(3)>,
  WATERMARK FOR f2.f22 AS f2.f22
)
```
will throw `ValidationException` because the schema resolver does not take type into account when validating the rowtime column.
https://github.com/apache/flink/blob/master/flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/catalog/DefaultSchemaResolver.java#L250-L259